netbsd/NetBSDProcessTable.c

```diff
@@ -182,7 +182,7 @@ static void NetBSDProcessTable_scanProcs(NetBSDProcessTable* this) {
          ProcessTable_add(&this->super, proc);
 
          proc->tty_nr = kproc->p_tdev;
-         const char* name = ((dev_t)kproc->p_tdev != KERN_PROC_TTY_NODEV) ? devname(kproc->p_tdev, S_IFCHR) : NULL;
```

Comment:
```c
(dev_t)kproc->p_tdev != KERN_PROC_TTY_NODEV)
```

There was a compiler warning stating this expression always evaluates to true. so removed it. The new build seems to be working fine after this removal.

Reply:
I dug the NetBSD header files, and this looks like a bug in the conditional. It probably meant to work like this:

```c
         const char* name = ((int64_t)kproc->p_tdev != (int64_t)KERN_PROC_TTY_NODEV) ? devname(kproc->p_tdev, S_IFCHR) : NULL;
```

Better make this change a separate commit as this one is a bug fix rather than a build error fix.

Reply:
All relevant definitions in NetBSD headers:

```c
// NetBSD <sys/types.h>
typedef uint64_t dev_t;
// NetBSD <sys/param.h>
#define NODEV (dev_t)(-1)
// NetBSD <sys/sysctl.h>
#define KERN_PROC_TTY_NODEV NODEV	
struct kinfo_proc2 {
...
uint32_t p_tdev;
...
};
```

Reply:
I had attempted this change when I was fixing the build and it still did not resolve the issue

```
netbsd/NetBSDProcessTable.c: In function 'NetBSDProcessTable_scanProcs':
netbsd/NetBSDProcessTable.c:185:53: warning: comparison is always true due to limited range of data type [-Wtype-limits]
          const char* name = ((int64_t)kproc->p_tdev != (int64_t)KERN_PROC_TTY_NODEV) ? devname(kproc->p_tdev, S_IFCHR): NULL;
```

I have reverted the change for now, if it causes trouble in the future will revisit it.

Reply:
@fraggerfox I made a typo. `uint32_t` cast to `int64_t` always results in a positive value.

Now I'm troubled. While the downcast like the following code would work, I really don't want to depend on the type of 
`(struct kinfo_proc2).p_tdev` for portability reason.

```c
         const char* name = (kproc->p_tdev != (uint32_t)KERN_PROC_TTY_NODEV) ? devname(kproc->p_tdev, S_IFCHR) : NULL;
```

Since I am not the author or maintainer of htop NetBSD port. I would stop here and let someone who is more familiar with NetBSD code to look at this.

Reply:
@Explorer09 : I agree with you, I think I need to look a bit deeper into the meaning of this check and why I put there in the first place. For now I see no harm in leaving it as it is.

P.S: I authored the [NetBSD](https://github.com/htop-dev/htop/pull/603) implementation a while back and I currently maintain it.

Reply:
@fraggerfox 
I finally figured out what to do. :)
The purpose of the code is to not allocate the name of the controlling TTY for processes that don't have one. (NetBSD source code shows that they do use `NODEV` placeholder value for `kproc->p_tdev`.) Although it might be safe to call [devname(3)](https://man.netbsd.org/devname.3) for a "no TTY device" placeholder. The `devname` function would return a placeholder string "??" which would not be ideal for displaying in htop. Hence the filter.

As `KERN_PROC_TTY_NODEV` is a negative constant. Compare with `~(KERN_PROC_TTY_NODEV)` would get rid of the signed and unsigned problem.

```c
         // KERN_PROC_TTY_NODEV is a negative constant but the type of 
         // kproc->p_tdev may be unsigned. 
         const char* name = ((dev_t)~kproc->p_tdev != (dev_t)~(KERN_PROC_TTY_NODEV)) ? devname(kproc->p_tdev, S_IFCHR) : NULL;
```

Reply:
@Explorer09 : Thank you for doing that investigation, shall I include your suggestion into this PR? (Now I remember about the `??` characters when I was doing the port originally)

I did a build locally and it seems to have worked!

Reply:
@fraggerfox You can include it, but as a separate commit. This way if another BSD port of htop has the problem people can adapt the patch easily.

Reply:
Agreed, I shall do a follow up PR once this is merged.